codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,328 @@
+    public void updateStreamableBytes(StreamState state) {
+        state(state.stream()).updateStreamableBytes(state.streamableBytes(),
+                                                    state.hasFrame() && state.isWriteAllowed());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // As long as there is some active frame we should write at least 1 time.
+        if (connectionState.isActiveCountForTree == 0) {
+            return false;
+        }
+
+        // The goal is to write until we write all the allocated bytes or are no longer making progress.
+        // We still attempt to write even after the number of allocated bytes has been exhausted to allow empty frames
+        // to be sent. Making progress means the active streams rooted at the connection stream has changed.
+        int oldIsActiveCountForTree;
+        do {
+            oldIsActiveCountForTree = connectionState.isActiveCountForTree;
+            // connectionState will never be active, so go right to its children.
+            maxBytes -= distributeToChildren(maxBytes, writer, connectionState);
+        } while (connectionState.isActiveCountForTree != 0 &&
+                (maxBytes > 0 || oldIsActiveCountForTree != connectionState.isActiveCountForTree));
+
+        return connectionState.isActiveCountForTree != 0;
+    }
+
+    /**
+     * Sets the amount of bytes that will be allocated to each stream. Defaults to 1KiB.
+     * @param minAllocationChunk the amount of bytes that will be allocated to each stream. Must be > 0.
+     */
+    public void allocationQuantum(int allocationQuantum) {
+        if (allocationQuantum <= 0) {
+            throw new IllegalArgumentException("allocationQuantum must be > 0");
+        }
+        this.allocationQuantum = allocationQuantum;
+    }
+
+    private int distribute(int maxBytes, Writer writer, State state) throws Http2Exception {
+        if (state.isActive) {
+            int nsent = min(maxBytes, state.streamableBytes);
+            state.write(nsent, writer);
+            if (nsent == 0 && maxBytes != 0) {
+                // If a stream sends zero bytes, then we gave it a chance to write empty frames and it is now
+                // considered inactive until the next call to updateStreamableBytes. This allows descendant streams to
+                // be allocated bytes when the parent stream can't utilize them. This may be as a result of the
+                // stream's flow control window being 0.
+                state.updateStreamableBytes(state.streamableBytes, false);
+            }
+            return nsent;
+        }
+
+        return distributeToChildren(maxBytes, writer, state);
+    }
+
+    private int distributeToChildren(int maxBytes, Writer writer, State state) throws Http2Exception {
+        long oldTotalQueuedWeights = state.totalQueuedWeights;
+        State childState = state.poll();
```

Comment:
I think we good the way things are...let me know if I'm missing something...

For the recursive case:
If a stream has no children (in the allocation tree) than that node must be active or it will not be in the allocation tree. If a node is active then it will not delegate to children and recursion ends.

For the initial case:
We check `connectionState.activeCountForTree == 0` before any allocation is done. So if the connection stream has no active children we don't get into this method.